cmd/iam-k8s-store.go

```diff
@@ -0,0 +1,554 @@
+		", ConfigMap: " + globalK8sIamStoreConfig.ConfigMapName)
+	return k8sStore
+}
+
+func (iamK8s *IAMK8sStore) ensureConfigMapExists() {
+	var objectMeta = metav1.ObjectMeta{Name: iamK8s.configMapName, Namespace: iamK8s.namespace}
+	var configMap = &corev1.ConfigMap{ObjectMeta: objectMeta}
+	_, err := iamK8s.configMapsClient.Create(context.Background(), configMap, metav1.CreateOptions{})
+	if err != nil && !errorsv1.IsAlreadyExists(err) {
+		panic(err)
+	}
+}
+
+func (iamK8s *IAMK8sStore) lock() {
+	iamK8s.Lock()
+}
+
+func (iamK8s *IAMK8sStore) unlock() {
+	iamK8s.Unlock()
+}
+
+func (iamK8s *IAMK8sStore) rlock() {
+	iamK8s.RLock()
+}
+
+func (iamK8s *IAMK8sStore) runlock() {
+	iamK8s.RUnlock()
+}
+
+// Should be called under config migration lock
+func (iamK8s *IAMK8sStore) migrateBackendFormat(ctx context.Context) error {
+	return nil
+}
+
+func isRetryable(err error) bool {
+	return errorsv1.IsTooManyRequests(err) || errorsv1.IsServiceUnavailable(err) || errorsv1.IsServerTimeout(err)
+}
+
+func backoff(attempt int, backoffTimeMs float64) {
+	durationMs := math.Pow(2, float64(attempt)) * backoffTimeMs
+	withJitter := rand.Intn(int(durationMs))
+	time.Sleep(time.Duration(withJitter) * time.Millisecond)
+}
+
+func backoffDefault(attempt int) {
+	backoff(attempt, 100)
+}
+
+func objPathToK8sValid(objPath string) string {
+	objPath = strings.TrimSuffix(objPath, ".json")
+	return strings.Replace(objPath, "/", ".", -1)
+}
+
+func objPathFromK8sValid(objPath string) string {
+	return strings.Replace(objPath, ".", "/", -1)
+}
+
+func toTtlExpKey(objPath string) string {
+	return "ttlExp_" + objPath
+}
+
+func trimTtlPrefix(ttlKey string) string {
+	return strings.TrimPrefix(ttlKey, "ttlExp_")
+}
+
+func hasTtlPrefix(ttlKey string) bool {
+	return strings.HasPrefix(ttlKey, "ttlExp_")
+}
+
+func (iamK8s *IAMK8sStore) saveIAMConfig(ctx context.Context, item interface{}, objPath string, opts ...options) error {
+	data, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	attempts := 0
+	for attempts < iamK8s.maxUpdateAttempts {
+		configMap, err := iamK8s.configMapsClient.Get(ctx, iamK8s.configMapName, metav1.GetOptions{})
+		if err != nil {
+			if isRetryable(err) {
+				attempts += 1
+				continue
+			} else {
+				return err
+			}
+		}
+		annotations := make(map[string]string)
+		if configMap.Annotations != nil {
+			annotations = configMap.Annotations
+		}
+		objPath := objPathToK8sValid(objPath)
+		annotations[objPath] = string(data)
+		if len(opts) > 0 {
+			annotations[toTtlExpKey(objPath)] = strconv.FormatInt(time.Now().Unix()+opts[0].ttl, 10)
+		}
+		configMapUpdated := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       iamK8s.namespace,
+				Name:            iamK8s.configMapName,
+				ResourceVersion: configMap.ResourceVersion,
+				Annotations:     annotations,
+			},
+		}
+		_, err = iamK8s.configMapsClient.Update(ctx, configMapUpdated, metav1.UpdateOptions{})
+		if err != nil {
```

Comment:
I wonder if the k8s error module uses error.As which returns false when err is nil. You also have a very similar err check in a later function, I would suggest merging them / re-use the logic if it's possible. But yes, I don't think you need to wrap this in a "err != nil", but I can't guarantee it.